src/verifier/az_snp_vtpm/mod.rs

```diff
@@ -0,0 +1,140 @@
+            .context("Failed to deserialize vTPM SEV-SNP evidence")?;
+
+        let hcl_report: HclReportWithRuntimeData = evidence.report[..].try_into()?;
+        let hashed_quote = hash_quote(attestation, &nonce);
+
+        verify_quote(&evidence.quote, &hcl_report, &hashed_quote)?;
+        verify_snp_report(&hcl_report)?;
+        verify_report_data(&hcl_report, &evidence.report)?;
+
+        let claim = parse_tee_evidence(&hcl_report);
+        Ok(claim)
+    }
+}
+
+fn verify_report_data(report: &HclReportWithRuntimeData, bytes: &[u8]) -> Result<()> {
+    let (_, var_data) = buf_to_hcl_data(bytes)?;
+    let var_data_hash = sha256(var_data);
+
+    // Only the first 32 bytes of SNP report data are used for the sha256
+    let report_data = &report.snp_report().report_data[..32];
+
+    if var_data_hash != report_data {
+        return Err(anyhow!(
+            "SNP report data field is not matching runtime data"
+        ));
+    }
+
+    Ok(())
+}
+
+fn verify_quote(
+    quote: &Quote,
+    report: &HclReportWithRuntimeData,
+    hashed_nonce: &[u8],
+) -> Result<()> {
+    let ak_pub = report.get_attestation_key()?;
+
+    let result = ak_pub
+        .verify_quote(quote, Some(hashed_nonce))
+        .context("Failed to verify vTPM quote")?;
+
+    if !result {
+        return Err(anyhow!("vTPM quote is invalid"));
+    }
+    Ok(())
+}
+
+fn parse_tee_evidence(hcl_report: &HclReportWithRuntimeData) -> TeeEvidenceParsedClaim {
+    let report = hcl_report.snp_report();
+    let numbers = [
+        ("policy_abi_major", report.policy.abi_major()),
+        ("policy_abi_minor", report.policy.abi_minor()),
+        ("policy_smt_allowed", report.policy.smt_allowed()),
+        ("policy_migrate_ma", report.policy.migrate_ma_allowed()),
+        ("policy_debug_allowed", report.policy.debug_allowed()),
+        (
+            "policy_single_socket",
+            report.policy.single_socket_required(),
+        ),
+        // versioning info
+        (
+            "reported_tcb_bootloader",
+            report.reported_tcb.boot_loader as u64,
+        ),
+        ("reported_tcb_tee", report.reported_tcb.tee as u64),
+        ("reported_tcb_snp", report.reported_tcb.snp as u64),
+        (
+            "reported_tcb_microcode",
+            report.reported_tcb.microcode as u64,
+        ),
+        // platform info
+        ("platform_tsme_enabled", report.plat_info.tsme_enabled()),
+        ("platform_smt_enabled", report.plat_info.smt_enabled()),
+    ];
+
+    let mut string_map: BTreeMap<_, _> = numbers.iter().map(|(k, v)| (*k, v.to_string())).collect();
+    string_map.insert("measurement", base64::encode(report.measurement));
+
+    json!(string_map) as TeeEvidenceParsedClaim
+}
+
+fn verify_snp_report(report: &HclReportWithRuntimeData) -> Result<()> {
+    let snp_report = report.snp_report();
+    let vcek = get_vcek_from_amd(snp_report)?;
+    let cert_chain = get_chain_from_amd()?;
```

Comment:
Do we plan to add caching for these?

Reply:
yes, that would make sense. at least parts of it.

Reply:
we'll bundle the certs (will use the milan pems from the SNP verifier, once it's merged)